### PR TITLE
Rewrite the lexer/parser to parse in more detail

### LINKED ISSUE #27: Gcode parser will remove comments, how can we keep them?
As I am still writing my gcode processor, I am finding out that after parsing and translating gcode all the comments of the initial gcode file are gone. I guess this was wanted because of the `fn skip_comments` in the lexer.

How could I keep the comments? Adding a new entry to `Gcode` seems not appropriate because comments do not have words, etc. However I think parsing the comments may be useful for this usage. I would also like to make the preprocessor add its own comments into the Gcode but it seems impossible either by using the typesystem, I would have to `println!` the comments to stdout directly.

My project can be found here if you need  a concrete example of what I need: https://git.microjoe.org/MicroJoe/emilly/

### PR DESCRIPTION
**Pull request by Michael-F-Bryan**
At the moment we throw a lot of information away meaning there's currently no way to inspect comments, tell whether two commands are part of the same block, or see the `%` sign (start/end of program marker). 

I'd also like to make it easier to create or transform gcodes, instead of just focusing on parsing them. There's no reason why this library should be limited to the motion controller executing gcodes, so making creation and transformation easy will let people write their own gcode generators (e.g. a slicer for a 3D printer) in Rust.

This PR essentially rewrites the library from the ground up to make the lexer and parser more comprehesive. It'd also be cool if we had a way to pretty-print or round-trip parsed/generated gcodes.

(fixes #27)

### CODE REVIEW

**Comment by Michael-F-Bryan** [comment]
@MicroJoe, would you like to have a look at this? The parser is still quite rough, but now each line will be parsed into a `Block` which contains zero or more comments and commands.

File: gcode/src/lexer.rs [hunk 1]
```diff
@@ -0,0 +1,394 @@
...
+                assert_eq!(token, should_be);
+                $span_check;
+            }
+        };
+    }
+
+    lexer_test!(lex_a_letter, "W" => Token::Letter('W'));
+    lexer_test!(lex_a_lowercase_letter, "g" => 'g');
+    lexer_test!(lex_comment_in_parens, "(this is a comment)" => "(this is a comment)");
+    lexer_test!(ignore_span lex_newline_comment, "; this is a comment\n" =>"; this is a comment");
```
Comment (MicroJoe) [inline gcode/src/lexer.rs]: missing space after operator

**Comment by Michael-F-Bryan** [comment]
> I have trouble representing myself what a "block" is, can you write a test scenario where we can see how blocks, comments and gcode works together from some "raw" gcode text?

A `Block` is just gcode's name for a line. It can contain multiple commands (and comments) that the gcode interpreter will execute at the same time.

[G-Code Basics: Program Format and Structure] is a pretty good resource for understanding how gcode programs are typically structured.


[G-Code Basics: Program Format and Structure]: https://www.cnccookbook.com/g-code-basics-program-format-structure-blocks/

File: gcode/src/lexer.rs (hunk 1 above)
Comment (Michael-F-Bryan) [reply gcode/src/lexer.rs]: What test case are you thinking of?

**Comment by MicroJoe** [comment]
I am thinking of a test that shows how a line like

```
G0 X0 Y0 ;possible comment?
```

would be parsed into rust (a block, containing a Gcode, and a Comment, that both are checked using assert_eq)

**Comment by Michael-F-Bryan** [comment]
@MicroJoe, good point. I'll rewrite [this test] to directly construct the expected value and use `assert_eq!()` instead of using the various getters.

[this test]: https://github.com/Michael-F-Bryan/gcode-rs/blob/0ddde7fde6bb6cef9a0e3a0b3c79a58dd95a7431/gcode/src/parser.rs#L441-L466

**Comment by Michael-F-Bryan** [comment]
We'll probably want some sort of builder type for creating new `Gcode`s or mutating existing ones, but this PR is a pretty solid start.

**Pull request merged**